trieste/bayesian_optimizer.py

```diff
@@ -39,47 +40,65 @@
 
 
 @dataclass(frozen=True)
-class LoggingState(Generic[S]):
-    """
-    Container for the state of the optimization process in :class:`BayesianOptimizer` at each
-    optimization step.
-    """
+class Record(Generic[S]):
+    """ Container to record the state of each step of the optimization process. """
 
     datasets: Mapping[str, Dataset]
-    """ All observer data at this optimization step. """
+    """ The known data from the observer. """
 
     models: Mapping[str, TrainableProbabilisticModel]
     """ The models over the :attr:`datasets`. """
 
     acquisition_state: Optional[S]
-    """ The acquisition state after this optimization step. """
+    """ The acquisition state. """
 
 
+# this should be a generic NamedTuple, but mypy doesn't support them
+#  https://github.com/python/mypy/issues/685
 @dataclass(frozen=True)
 class OptimizationResult(Generic[S]):
-    """ Container for the final result of the optimization process. """
+    """ The final result, and the historical data of the optimization process. """
 
-    datasets: Mapping[str, Dataset]
+    final_result: Result[Record[S]]
     """
-    All data from the observer (unless :attr:`error` is populated, in which case this is the data
-    from the point at which the process was interrupted).
+    The final result of the optimization process. This contains either a :class:`Record` or an
+    exception.
     """
 
-    models: Mapping[str, TrainableProbabilisticModel]
-    """
-    The models over the :attr:`datasets` (unless :attr:`error` is populated, in which case this is
-    the models from the point at which the process was interrupted). """
+    history: List[Record[S]]
+    r""" The history of the :class:`Record`\ s from each step of the optimization process. """
 
-    history: List[LoggingState[S]]
-    """ The data, models, and acquisition state at each completed optimization step. """
+    def astuple(self) -> Tuple[Result[Record[S]], List[Record[S]]]:
+        """
+        **Note:** In contrast to the standard library function :func:`dataclasses.astuple`, this
+        method does *not* deepcopy instance attributes.
 
-    error: Optional[Exception]
-    """ The exception that occurred, if any. """
+        :return: The :attr:`final_result` and :attr:`history` as a 2-tuple.
+        """
+        return self.final_result, self.history
+
+    def try_final_data(self) -> Mapping[str, Dataset]:
```

Comment:
`try_get_final_data` might make it clearer what this does, even if it is slightly longer? I was a bit confused the first time I read this, I initially thought it was something to do with getting model predictions for the most recent data.

Reply:
hmmm yes i see what you mean. The name never quite sat right

Reply:
done